### PR TITLE
[SPARK-40142][PYTHON][SQL][FOLLOW-UP] Make pyspark.sql.functions examples self-contained (part 2, 32 functions)

### PR DESCRIPTION
**Pull request by khalidmammadov**
### What changes were proposed in this pull request?
Docstring improvements

### Why are the changes needed?
To help users to understand pyspark API

### Does this PR introduce _any_ user-facing change?
Yes, documentation

### How was this patch tested?
`bundle exec jekyll serve --host 0.0.0.0`

### CODE REVIEW

File: python/pyspark/sql/functions.py [hunk 1]
```diff
@@ -994,12 +1014,22 @@ def cot(col: "ColumnOrName") -> Column:
     Parameters
     ----------
     col : :class:`~pyspark.sql.Column` or str
-        Angle in radians
+        angle in radians.
```
Comment (srowen) [inline python/pyspark/sql/functions.py:1017]: Why these changes? seemed better before

File: python/pyspark/sql/functions.py [hunk 2]
```diff
@@ -1142,13 +1384,23 @@ def sinh(col: "ColumnOrName") -> Column:
     Parameters
     ----------
     col : :class:`~pyspark.sql.Column` or str
-        hyperbolic angle
+        hyperbolic angle in radians
```
Comment (srowen) [inline python/pyspark/sql/functions.py:1387]: I dont' believe hyperbolic angles are measured in radians

**Review by srowen — CHANGES_REQUESTED** [review]

File: python/pyspark/sql/functions.py [hunk 3]
```diff
@@ -6158,12 +6764,14 @@ def _test() -> None:
...
     from pyspark.sql import Row, SparkSession
     import pyspark.sql.functions
+    import math
 
     globs = pyspark.sql.functions.__dict__.copy()
     spark = SparkSession.builder.master("local[4]").appName("sql.functions tests").getOrCreate()
     sc = spark.sparkContext
     globs["sc"] = sc
     globs["spark"] = spark
+    globs["math"] = math
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:6774]: Let's import this within the docstring. I think we should ideally remove all these globals (except `spark`).

**Comment by HyukjinKwon** [comment]
Thanks for doing this, @khalidmammadov

File: python/pyspark/sql/functions.py (hunk 1 above)
Comment (khalidmammadov) [reply python/pyspark/sql/functions.py:1010]: There are different conventions in this file and most prominent ones is lower case and dot at the end. Tried to be consistent. Can revert if needed

File: python/pyspark/sql/functions.py (hunk 2 above)
Comment (khalidmammadov) [reply python/pyspark/sql/functions.py]: my bad, reverted

File: python/pyspark/sql/functions.py (hunk 3 above)
Comment (khalidmammadov) [reply python/pyspark/sql/functions.py]: removed, I can look into removal of "sc" in the next PRs

**Comment by khalidmammadov** [comment]
> Thanks for doing this, @khalidmammadov

my pleasure

File: python/pyspark/sql/functions.py [hunk 4]
```diff
@@ -1037,6 +1084,22 @@ def expm1(col: "ColumnOrName") -> Column:
...
+    Returns
+    -------
+    :class:`~pyspark.sql.Column`
+        exponential less one.
+
+    Examples
+    --------
+    >>> df = spark.range(1)
+    >>> df.select(expm1(lit(1))).first()  # doctest: +ELLIPSIS
+    Row(EXPM1(1)=1.71828...)
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:1102]: This is a nice trick.

File: python/pyspark/sql/functions.py [hunk 5]
```diff
@@ -1343,6 +1798,26 @@ def var_pop(col: "ColumnOrName") -> Column:
...
+        variance of given column.
+
+    Examples
+    --------
+    >>> df = spark.range(6)
+    >>> df.select(var_pop(df.id)).show()
+    +------------------+
+    |       var_pop(id)|
+    +------------------+
+    |2.9166666666666665|
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:1819]: I think we should probably do ELLIPSIS since float representation is flaky in Python (and JDK 11/17 IIRC)

File: python/pyspark/sql/functions.py [hunk 6]
```diff
@@ -1352,6 +1827,26 @@ def skewness(col: "ColumnOrName") -> Column:
...
+        skewness of given column.
+
+    Examples
+    --------
+    >>> df = spark.createDataFrame([[1],[1],[2]], ["c"])
+    >>> df.select(skewness(df.c)).show()
+    +------------------+
+    |       skewness(c)|
+    +------------------+
+    |0.7071067811865475|
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:1848]: ditto

File: python/pyspark/sql/functions.py [hunk 7]
```diff
@@ -1315,6 +1710,26 @@ def stddev_pop(col: "ColumnOrName") -> Column:
...
+        standard deviation of given column.
+
+    Examples
+    --------
+    >>> df = spark.range(6)
+    >>> df.select(stddev_pop(df.id)).show()
+    +-----------------+
+    |   stddev_pop(id)|
+    +-----------------+
+    |1.707825127659933|
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:1731]: ditto

File: python/pyspark/sql/functions.py [hunk 8]
```diff
@@ -1305,6 +1680,26 @@ def stddev_samp(col: "ColumnOrName") -> Column:
...
+        standard deviation of given column.
+
+    Examples
+    --------
+    >>> df = spark.range(6)
+    >>> df.select(stddev_samp(df.id)).show()
+    +------------------+
+    |   stddev_samp(id)|
+    +------------------+
+    |1.8708286933869707|
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:1701]: ditto

File: python/pyspark/sql/functions.py [hunk 9]
```diff
@@ -1295,6 +1650,26 @@ def stddev(col: "ColumnOrName") -> Column:
...
+        standard deviation of given column.
+
+    Examples
+    --------
+    >>> df = spark.range(6)
+    >>> df.select(stddev(df.id)).show()
+    +------------------+
+    |   stddev_samp(id)|
+    +------------------+
+    |1.8708286933869707|
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:1671]: ditto

**Review by HyukjinKwon — APPROVED** [review]
LGTM otherwise

File: python/pyspark/sql/functions.py [hunk 10]
```diff
@@ -962,6 +962,13 @@ def cos(col: "ColumnOrName") -> Column:
     -------
     :class:`~pyspark.sql.Column`
         cosine of the angle, as if computed by `java.lang.Math.cos()`.
+
+    Examples
+    --------
+    >>> import math
+    >>> df = spark.range(1)
+    >>> df.select(cos(lit(math.pi))).first()  # doctest: +ELLIPSIS
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:970]:
```suggestion
    >>> df.select(cos(lit(math.pi))).first()
```

File: python/pyspark/sql/functions.py [hunk 11]
```diff
@@ -981,6 +988,12 @@ def cosh(col: "ColumnOrName") -> Column:
     -------
     :class:`~pyspark.sql.Column`
         hyperbolic cosine of the angle, as if computed by `java.lang.Math.cosh()`
+
+    Examples
+    --------
+    >>> df = spark.range(1)
+    >>> df.select(cosh(lit(1))).first()  # doctest: +ELLIPSIS
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:995]:
```suggestion
    >>> df.select(cosh(lit(1))).first()
```

File: python/pyspark/sql/functions.py [hunk 12]
```diff
@@ -994,12 +1007,19 @@ def cot(col: "ColumnOrName") -> Column:
...
     -------
     :class:`~pyspark.sql.Column`
-        Cotangent of the angle.
+        cotangent of the angle.
+
+    Examples
+    --------
+    >>> import math
+    >>> df = spark.range(1)
+    >>> df.select(cot(lit(math.radians(45)))).first()  # doctest: +ELLIPSIS
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:1021]:
```suggestion
    >>> df.select(cot(lit(math.radians(45)))).first()
```

File: python/pyspark/sql/functions.py [hunk 13]
```diff
@@ -1013,12 +1033,19 @@ def csc(col: "ColumnOrName") -> Column:
...
     -------
     :class:`~pyspark.sql.Column`
-        Cosecant of the angle.
+        cosecant of the angle.
+
+    Examples
+    --------
+    >>> import math
+    >>> df = spark.range(1)
+    >>> df.select(csc(lit(math.radians(90)))).first()  # doctest: +ELLIPSIS
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:1047]:
```suggestion
    >>> df.select(csc(lit(math.radians(90)))).first()
```

File: python/pyspark/sql/functions.py [hunk 14]
```diff
@@ -1037,6 +1084,22 @@ def expm1(col: "ColumnOrName") -> Column:
...
+
+    Returns
+    -------
+    :class:`~pyspark.sql.Column`
+        exponential less one.
+
+    Examples
+    --------
+    >>> df = spark.range(1)
+    >>> df.select(expm1(lit(1))).first()  # doctest: +ELLIPSIS
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:1101]:
```suggestion
    >>> df.select(expm1(lit(1))).first()
```

File: python/pyspark/sql/functions.py [hunk 15]
```diff
@@ -1055,6 +1138,23 @@ def log(col: "ColumnOrName") -> Column:
...
+    Returns
+    -------
+    :class:`~pyspark.sql.Column`
+        natural logarithm of the given value.
+
+    Examples
+    --------
+    >>> import math
+    >>> df = spark.range(1)
+    >>> df.select(log(lit(math.e))).first()  # doctest: +ELLIPSIS
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:1156]:
```suggestion
    >>> df.select(log(lit(math.e))).first()
```

File: python/pyspark/sql/functions.py [hunk 16]
```diff
@@ -1064,15 +1164,57 @@ def log10(col: "ColumnOrName") -> Column:
...
+    Returns
+    -------
+    :class:`~pyspark.sql.Column`
+        natural logarithm of the "given value plus one".
+
+    Examples
+    --------
+    >>> import math
+    >>> df = spark.range(1)
+    >>> df.select(log1p(lit(math.e))).first()  # doctest: +ELLIPSIS
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:1211]:
```suggestion
    >>> df.select(log1p(lit(math.e))).first()
```

File: python/pyspark/sql/functions.py [hunk 17]
```diff
@@ -1064,15 +1164,57 @@ def log10(col: "ColumnOrName") -> Column:
...
+    Row(LOG1P(2.71828...)=1.31326...)
+
+    Same as:
+
+    >>> df.select(log(lit(math.e+1))).first()  # doctest: +ELLIPSIS
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:1216]:
```suggestion
    >>> df.select(log(lit(math.e+1))).first()
```

File: python/pyspark/sql/functions.py [hunk 18]
```diff
@@ -1102,6 +1271,12 @@ def sec(col: "ColumnOrName") -> Column:
     -------
     :class:`~pyspark.sql.Column`
         Secant of the angle.
+
+    Examples
+    --------
+    >>> df = spark.range(1)
+    >>> df.select(sec(lit(1.5))).first()  # doctest: +ELLIPSIS
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:1278]:
```suggestion
    >>> df.select(sec(lit(1.5))).first()
```

File: python/pyspark/sql/functions.py [hunk 19]
```diff
@@ -1124,11 +1326,19 @@ def sin(col: "ColumnOrName") -> Column:
...
     Returns
     -------
     :class:`~pyspark.sql.Column`
         sine of the angle, as if computed by `java.lang.Math.sin()`
+
+    Examples
+    --------
+    >>> import math
+    >>> df = spark.range(1)
+    >>> df.select(sin(lit(math.radians(90)))).first()  # doctest: +ELLIPSIS
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:1340]:
```suggestion
    >>> df.select(sin(lit(math.radians(90)))).first()
```

File: python/pyspark/sql/functions.py [hunk 20]
```diff
@@ -1142,13 +1352,19 @@ def sinh(col: "ColumnOrName") -> Column:
...
     Returns
     -------
     :class:`~pyspark.sql.Column`
         hyperbolic sine of the given value,
         as if computed by `java.lang.Math.sinh()`
+
+    Examples
+    --------
+    >>> df = spark.range(1)
+    >>> df.select(sinh(lit(1.1))).first()  # doctest: +ELLIPSIS
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:1366]:
```suggestion
    >>> df.select(sinh(lit(1.1))).first()
```

File: python/pyspark/sql/functions.py [hunk 21]
```diff
@@ -1168,6 +1384,13 @@ def tan(col: "ColumnOrName") -> Column:
     -------
     :class:`~pyspark.sql.Column`
         tangent of the given value, as if computed by `java.lang.Math.tan()`
+
+    Examples
+    --------
+    >>> import math
+    >>> df = spark.range(1)
+    >>> df.select(tan(lit(math.radians(45)))).first()  # doctest: +ELLIPSIS
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:1392]:
```suggestion
    >>> df.select(tan(lit(math.radians(45)))).first()
```

File: python/pyspark/sql/functions.py [hunk 22]
```diff
@@ -1188,6 +1411,13 @@ def tanh(col: "ColumnOrName") -> Column:
     :class:`~pyspark.sql.Column`
         hyperbolic tangent of the given value
         as if computed by `java.lang.Math.tanh()`
+
+    Examples
+    --------
+    >>> import math
+    >>> df = spark.range(1)
+    >>> df.select(tanh(lit(math.radians(90)))).first()  # doctest: +ELLIPSIS
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:1419]:
```suggestion
    >>> df.select(tanh(lit(math.radians(90)))).first()
```

File: python/pyspark/sql/functions.py [hunk 23]
```diff
@@ -1295,6 +1651,22 @@ def stddev(col: "ColumnOrName") -> Column:
...
+
+    Returns
+    -------
+    :class:`~pyspark.sql.Column`
+        standard deviation of given column.
+
+    Examples
+    --------
+    >>> df = spark.range(6)
+    >>> df.select(stddev(df.id)).first()  # doctest: +ELLIPSIS
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:1668]:
```suggestion
    >>> df.select(stddev(df.id)).first()
```

File: python/pyspark/sql/functions.py [hunk 24]
```diff
@@ -1305,6 +1677,22 @@ def stddev_samp(col: "ColumnOrName") -> Column:
...
+
+    Returns
+    -------
+    :class:`~pyspark.sql.Column`
+        standard deviation of given column.
+
+    Examples
+    --------
+    >>> df = spark.range(6)
+    >>> df.select(stddev_samp(df.id)).first()  # doctest: +ELLIPSIS
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:1694]:
```suggestion
    >>> df.select(stddev_samp(df.id)).first()
```

File: python/pyspark/sql/functions.py [hunk 25]
```diff
@@ -1315,6 +1703,22 @@ def stddev_pop(col: "ColumnOrName") -> Column:
...
+
+    Returns
+    -------
+    :class:`~pyspark.sql.Column`
+        standard deviation of given column.
+
+    Examples
+    --------
+    >>> df = spark.range(6)
+    >>> df.select(stddev_pop(df.id)).first()  # doctest: +ELLIPSIS
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:1720]:
```suggestion
    >>> df.select(stddev_pop(df.id)).first()
```

File: python/pyspark/sql/functions.py [hunk 26]
```diff
@@ -1343,6 +1787,22 @@ def var_pop(col: "ColumnOrName") -> Column:
...
+
+    Returns
+    -------
+    :class:`~pyspark.sql.Column`
+        variance of given column.
+
+    Examples
+    --------
+    >>> df = spark.range(6)
+    >>> df.select(var_pop(df.id)).first()  # doctest: +ELLIPSIS
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:1804]:
```suggestion
    >>> df.select(var_pop(df.id)).first()
```

File: python/pyspark/sql/functions.py [hunk 27]
```diff
@@ -1352,6 +1812,22 @@ def skewness(col: "ColumnOrName") -> Column:
...
+
+    Returns
+    -------
+    :class:`~pyspark.sql.Column`
+        skewness of given column.
+
+    Examples
+    --------
+    >>> df = spark.createDataFrame([[1],[1],[2]], ["c"])
+    >>> df.select(skewness(df.c)).first()  # doctest: +ELLIPSIS
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:1829]:
```suggestion
    >>> df.select(skewness(df.c)).first()
```

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by HyukjinKwon** [comment]
Thanks for working on this @khalidmammadov

**Comment by khalidmammadov** [comment]
> Thanks for working on this @khalidmammadov

No worries, I will pick up next batch then from functions
